bfvmm/src/vmcs/src/vmcs_intel_x64_debug.cpp

```diff
@@ -22,524 +23,716 @@
-    PRINT_CONTROL(VM_ENTRY_CONTROL_DEACTIVATE_DUAL_MONITOR_TREATMENT);
-    PRINT_CONTROL(VM_ENTRY_CONTROL_LOAD_IA32_PERF_GLOBAL_CTRL);
-    PRINT_CONTROL(VM_ENTRY_CONTROL_LOAD_IA32_PAT);
-    PRINT_CONTROL(VM_ENTRY_CONTROL_LOAD_IA32_EFER);
+    dump_vm_control(vm_entry_controls::load_debug_controls::name,
+                    vm_entry_controls::load_debug_controls::is_enabled(),
+                    ia32_vmx_true_entry_ctls::load_debug_controls::is_allowed0(),
+                    ia32_vmx_true_entry_ctls::load_debug_controls::is_allowed1());
+
+    dump_vm_control(vm_entry_controls::ia_32e_mode_guest::name,
+                    vm_entry_controls::ia_32e_mode_guest::is_enabled(),
+                    ia32_vmx_true_entry_ctls::ia_32e_mode_guest::is_allowed0(),
+                    ia32_vmx_true_entry_ctls::ia_32e_mode_guest::is_allowed1());
+
+    dump_vm_control(vm_entry_controls::entry_to_smm::name,
+                    vm_entry_controls::entry_to_smm::is_enabled(),
+                    ia32_vmx_true_entry_ctls::entry_to_smm::is_allowed0(),
+                    ia32_vmx_true_entry_ctls::entry_to_smm::is_allowed1());
+
+    dump_vm_control(vm_entry_controls::deactivate_dual_monitor_treatment::name,
+                    vm_entry_controls::deactivate_dual_monitor_treatment::is_enabled(),
+                    ia32_vmx_true_entry_ctls::deactivate_dual_monitor_treatment::is_allowed0(),
+                    ia32_vmx_true_entry_ctls::deactivate_dual_monitor_treatment::is_allowed1());
+
+    dump_vm_control(vm_entry_controls::load_ia32_perf_global_ctrl::name,
+                    vm_entry_controls::load_ia32_perf_global_ctrl::is_enabled(),
+                    ia32_vmx_true_entry_ctls::load_ia32_perf_global_ctrl::is_allowed0(),
+                    ia32_vmx_true_entry_ctls::load_ia32_perf_global_ctrl::is_allowed1());
+
+    dump_vm_control(vm_entry_controls::load_ia32_pat::name,
+                    vm_entry_controls::load_ia32_pat::is_enabled(),
+                    ia32_vmx_true_entry_ctls::load_ia32_pat::is_allowed0(),
+                    ia32_vmx_true_entry_ctls::load_ia32_pat::is_allowed1());
+
+    dump_vm_control(vm_entry_controls::load_ia32_efer::name,
+                    vm_entry_controls::load_ia32_efer::is_enabled(),
+                    ia32_vmx_true_entry_ctls::load_ia32_efer::is_allowed0(),
+                    ia32_vmx_true_entry_ctls::load_ia32_efer::is_allowed1());
 
     bfdebug << bfendl;
 }
 
-#endif
+void
+debug::dump_vmcs_field(vmcs::field_type addr, const char *name, bool exists)
+{
+    bfdebug << name << ": " ;
+
+    if (!exists)
+        bfinfo << "doesn't exist" << bfendl;
+    else
+        bfinfo << view_as_pointer(vm::read(addr, name)) << bfendl;
+}
+
+void
+debug::dump_vm_control(const char *name, bool is_set,
+                       bool is_allowed0, bool is_allowed1)
+{
+    bfdebug << name << ": ";
+
+    bfinfo << "value == " << is_set
+           << ", is_allowed0 == " << is_allowed0
```

Comment:
To simplify things for now... ditch the is_allowed stuff... and just format this like the flags fields I showed above. If the flag is enabled.... show it... otherwise don't. If they want to know what is allowed / not allowed, the current MSR dump code shows that.